File: frontend/src/types/experiment.ts
// 实验状态
export type ExperimentStatus = 'draft' | 'in_progress' | 'completed' | 'cancelled'

// 实验类型
export type ExperimentType = 'performance' | 'reliability' | 'environment' | 'material' | 'process' | 'other'

// 实验结论
export type RecordConclusion = 'pass' | 'fail' | 'conditional_pass' | 'need_retest'

// 参数模板项
export interface ParamTemplateItem {
  name: string
  unit?: string | null
  default?: any
  min?: number | null
  max?: number | null
}

export interface ExperimentOut {
  id: number
  code: string
  project_id: number
  name: string
  description?: string
  exp_type: ExperimentType
  status: ExperimentStatus
  designer_id?: number
  designer_name?: string
  executor_id?: number
  executor_name?: string
  plan_start?: string
  plan_end?: string
  actual_start?: string
  actual_end?: string
  param_template?: ParamTemplateItem[] | any
  created_at: string
  updated_at: string
}

export interface ExperimentAttachmentOut {
  id: number
  record_id: number
  file_name: string
  file_path: string
  file_size: number
  uploaded_at: string
}

export interface ExperimentRecordOut {
  id: number
  experiment_id: number
  batch_no?: string
  sample_code?: string
  executor_id?: number
  executor_name?: string
  param_values?: any
  result_data?: any
  result_summary?: string
  conclusion?: RecordConclusion
  is_abnormal: boolean
  abnormal_desc?: string
  executed_at?: string
  created_at: string
  attachment_count?: number
  attachments?: ExperimentAttachmentOut[]
}

// 状态/类型/结论下拉选项（与后端枚举保持一致）
export const EXPERIMENT_STATUS_OPTS = [
  { label: '草稿', value: 'draft', tag: 'info' as const },
  { label: '进行中', value: 'in_progress', tag: 'warning' as const },
  { label: '已完成', value: 'completed', tag: 'success' as const },
  { label: '已取消', value: 'cancelled', tag: 'danger' as const },
]

export const EXPERIMENT_TYPE_OPTS = [
  { label: '性能测试', value: 'performance' },
  { label: '可靠性测试', value: 'reliability' },
  { label: '环境测试', value: 'environment' },
  { label: '材料试验', value: 'material' },
  { label: '工艺验证', value: 'process' },
  { label: '其他', value: 'other' },
]

export const CONCLUSION_OPTS = [
  { label: '通过', value: 'pass', tag: 'success' as const },
  { label: '有条件通过', value: 'conditional_pass', tag: 'warning' as const },
  { label: '需复测', value: 'need_retest', tag: 'warning' as const },
  { label: '失败', value: 'fail', tag: 'danger' as const },
]

export const statusLabel = (s?: string) =>
  EXPERIMENT_STATUS_OPTS.find(o => o.value === s)?.label || s || '-'
export const statusTag = (s?: string) =>
  EXPERIMENT_STATUS_OPTS.find(o => o.value === s)?.tag || 'info'
export const typeLabel = (t?: string) =>
  EXPERIMENT_TYPE_OPTS.find(o => o.value === t)?.label || t || '-'
export const conclusionLabel = (c?: string) =>
  CONCLUSION_OPTS.find(o => o.value === c)?.label || c || '-'
export const conclusionTag = (c?: string) =>
  CONCLUSION_OPTS.find(o => o.value === c)?.tag || 'info'

File: backend/app/schemas/experiment.py
"""研发实验管理 Schema"""
from pydantic import BaseModel, Field
from datetime import datetime, date
from typing import Optional, List, Any
from enum import Enum
from app.schemas.common import PaginationParams


# ===== 枚举 =====
class ExperimentStatus(str, Enum):
    DRAFT = "draft"               # 草稿
    IN_PROGRESS = "in_progress"   # 进行中
    COMPLETED = "completed"       # 已完成
    CANCELLED = "cancelled"       # 已取消


class ExperimentType(str, Enum):
    PERFORMANCE = "performance"   # 性能测试
    RELIABILITY = "reliability"   # 可靠性测试
    ENVIRONMENT = "environment"   # 环境测试
    MATERIAL = "material"         # 材料试验
    PROCESS = "process"           # 工艺验证
    OTHER = "other"               # 其他


class RecordConclusion(str, Enum):
    PASS = "pass"                       # 通过
    FAIL = "fail"                       # 失败
    CONDITIONAL_PASS = "conditional_pass"  # 有条件通过
    NEED_RETEST = "need_retest"         # 需复测


# ===== 参数模板项 =====
class ParamTemplateItem(BaseModel):
    name: str = Field(..., description="参数名称")
    unit: Optional[str] = Field(None, description="单位")
    default: Optional[Any] = Field(None, description="默认值")
    min: Optional[float] = Field(None, description="下限")
    max: Optional[float] = Field(None, description="上限")


# ===== Experiment =====
class ExperimentBase(BaseModel):
    name: str = Field(..., max_length=200)
    description: Optional[str] = None
    exp_type: ExperimentType = ExperimentType.PERFORMANCE
    status: ExperimentStatus = ExperimentStatus.DRAFT
    designer_id: Optional[int] = None
    executor_id: Optional[int] = None
    plan_start: Optional[date] = None
    plan_end: Optional[date] = None
    param_template: Optional[List[ParamTemplateItem]] = None


class ExperimentCreate(ExperimentBase):
    project_id: int


class ExperimentUpdate(BaseModel):
    name: Optional[str] = Field(None, max_length=200)
    description: Optional[str] = None
    exp_type: Optional[ExperimentType] = None
    status: Optional[ExperimentStatus] = None
    executor_id: Optional[int] = None
    plan_start: Optional[date] = None
    plan_end: Optional[date] = None
    actual_start: Optional[date] = None
    actual_end: Optional[date] = None
    param_template: Optional[List[ParamTemplateItem]] = None


class ExperimentOut(ExperimentBase):
    id: int
    code: str
    project_id: int
    actual_start: Optional[date] = None
    actual_end: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ExperimentDetailOut(ExperimentOut):
    """实验详情（含展示字段）"""
    designer_name: Optional[str] = None
    executor_name: Optional[str] = None


# ===== ExperimentRecord =====
class ExperimentRecordBase(BaseModel):
    batch_no: Optional[str] = Field(None, max_length=50)
    sample_code: Optional[str] = Field(None, max_length=50)
    executor_id: Optional[int] = None
    param_values: Optional[dict] = None
    result_data: Optional[dict] = None
    result_summary: Optional[str] = None
    conclusion: Optional[RecordConclusion] = None
    is_abnormal: bool = False
    abnormal_desc: Optional[str] = None
    executed_at: Optional[date] = None


class ExperimentRecordCreate(ExperimentRecordBase):
    experiment_id: int


class ExperimentRecordUpdate(BaseModel):
    batch_no: Optional[str] = Field(None, max_length=50)
    sample_code: Optional[str] = Field(None, max_length=50)
    executor_id: Optional[int] = None
    param_values: Optional[dict] = None
    result_data: Optional[dict] = None
    result_summary: Optional[str] = None
    conclusion: Optional[RecordConclusion] = None
    is_abnormal: Optional[bool] = None
    abnormal_desc: Optional[str] = None
    executed_at: Optional[date] = None


class ExperimentRecordOut(ExperimentRecordBase):
    id: int
    experiment_id: int
    created_at: datetime

    class Config:
        from_attributes = True


class ExperimentRecordDetailOut(ExperimentRecordOut):
    """记录详情（含展示字段）"""
    executor_name: Optional[str] = None
    attachment_count: int = 0
    attachments: List[Any] = []  # List[ExperimentAttachmentOut]，延迟引用避免循环


# ===== 附件 =====
class ExperimentAttachmentOut(BaseModel):
    id: int
    record_id: int
    file_name: str
    file_path: str
    file_size: int
    uploaded_at: datetime

    class Config:
        from_attributes = True


# 修正 ExperimentRecordDetailOut.attachments 的类型
ExperimentRecordDetailOut.model_rebuild()


# ===== 状态流转动作 =====
class StatusAction(BaseModel):
    action: str = Field(..., description="start | complete | cancel")


# ===== 查询 =====
class ExperimentQuery(PaginationParams):
    project_id: Optional[int] = None
    status: Optional[ExperimentStatus] = None
    exp_type: Optional[ExperimentType] = None
    keyword: Optional[str] = None
    date_from: Optional[date] = None
    date_to: Optional[date] = None


class RecordQuery(PaginationParams):
    """实验记录查询"""
    batch_no: Optional[str] = None
    sample_code: Optional[str] = None
    conclusion: Optional[RecordConclusion] = None
    is_abnormal: Optional[bool] = None
    date_from: Optional[date] = None
    date_to: Optional[date] = None


class BatchDeleteRecords(BaseModel):
    """批量删除记录"""
    ids: List[int]
